### ISSUE TITLE
Program arguments are removed if wildcard match fails

### ISSUE THREAD

**Issue by jarun** [fish-shell/fish-shell#3365]
```
$ fish --version 
fish, version 2.2.0
$ uname -a
Linux tux 4.4.0-36-generic #55-Ubuntu SMP Thu Aug 11 18:01:55 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
Terminal emulator: [AltYo](https://github.com/linvinus/AltYo) (should happen with any)

Issue: While using my utility [imgd](https://github.com/jarun/imgd) I noticed that arguments are completely removed when there are no files matching a wildcard.

Example of an instance with `fish` when there are no files with `abc` extension in `~/Documents`:
```
$ imgd -x 1366x1000 -rac ~/Documents/*.abc
No matches for wildcard “~/Documents/*.abc”.
fish: imgd -x 1366x1000 -rac ~/Documents/*.abc
                             ^
usage: imgd [-h] [-s %] [-x HxV] [-o deg] [-a] [-c] [-d] [-e] [-f] [-i] [-k]
            [-n] [-p] [-q] [-r] [-w] [-z]
            PATH [PATH ...]
imgd: error: the following arguments are required: PATH
```
The error from argparse shows that `~/Documents/*.abc` is not passed to the program from the shell.
However, bash works as expected:
```
imgd -x 1366x1000 -rac ~/Documents/*.abc
/home/vaio/Documents/*.abc does not exist
```
i.e., the argument `~/Documents/*.abc` is passed to the program with bash. As per my understanding this should be the right behaviour as the argument was intended to be handled by the program, not the shell. And the operational behaviour when no match is found would vary from program to program.

**Comment by ridiculousfish** [comment]
What does `imgd` do when passed the literal `~/Documents/*.abc`?

**Comment by ridiculousfish** [comment]
This is one of those quite insane bash behaviors, where globs that fail to expand are passed literally to the command. It doesn't seem like `imgd` is doing anything useful with it, since it just prints that the filename does not exist.

Note that `imgd` is NOT attempt to expand the glob! It's looking for a literal file `*.abc`. To operate on such a literal file, you could pass `\*.abc` instead, which works in both bash and fish.

The bash behavior of globs that expand contextually (depending on the command and the filesystem state) is very confusing. It's also dangerous, since users may introduce a glob accidentally (e.g. `foo?bar`) which will work fine right up until it matches something. 

The behavior where globs expand to nothing was changed in fish 2.3.0 - see #2394. In fish 2.3.0 your imgd command will not execute, instead of being executed with zero arguments.

Thanks for filing this. This is one of those places where fish deliberately diverges from bash, and the justification is strong. Closing as not to be fixed.

**Comment by jarun** [comment]
I have also gone through issues #683, #796 and #799 and understood your stand in favour of this behaviour.
My problem is: `imgd` can take multiple source files. My intention is to process the current directory by default if no source is provided. I had t block the behavour in commit https://github.com/jarun/imgd/commit/1fd5e4aa789add432bd8514269dc4433b37ab203 because there is no way for `imgd` to know if ANY src files were at all specified as fish strips it.

**Comment by jarun** [comment]
All I am asking for is a way to know that an argument was passed.

**Comment by ridiculousfish** [comment]
My apologies, I didn't realize you were the author of `imgd`!

The behavior of fish 2.3.x (the most recent release) is that globs never expand to zero arguments; instead fish simply does not execute the command if a glob fails. This was fixed in #2394. Based on this, it seems like you can restore the behavior you want, since fish will never remove a glob argument outright.

**Comment by jarun** [comment]
Awesome! Just what I was looking for.

BTW, just to leave a note of it, I was also thinking of another solution keeping the current behaviour intact. When an wildcard is stripped, set an environment variable (similar to the standard `errno` stuff). Programs dealing in positional arguments can have an additional check for really `0` arguments and take appropriate action.

However, I think the current approach is much more elegant as it is also `backward compatible` ;).
